internal/test/e2e/tinkerbell_hardware_queue.go

```diff
@@ -0,0 +1,59 @@
+package e2e
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/aws/eks-anywhere/internal/pkg/api"
+)
+
+// Default timeout for Tink tests to poll for hardware.
+const hwPollingTimeout = 120 * time.Minute
+
+// hardwareCatalogue has a thread safe FIFO queue implementation to facilitate hardware reservation.
+type hardwareCatalogue struct {
+	hws []*api.Hardware
+	mu  sync.Mutex
+}
+
+func (hwQu *hardwareCatalogue) reserveHardware(count int) ([]*api.Hardware, error) {
+	now := time.Now()
+	after := now.Add(hwPollingTimeout)
+	for {
+		if now.After(after) {
+			return nil, fmt.Errorf("hardware polling request timed out")
+		}
+		hwQu.mu.Lock()
+		if count <= len(hwQu.hws) {
+			hardwareReserved := hwQu.releaseHw(count)
+			hwQu.mu.Unlock()
+			return hardwareReserved, nil
+		}
+		hwQu.mu.Unlock()
+		time.Sleep(1 * time.Minute)
+	}
+}
+
+func (hwQu *hardwareCatalogue) releaseHardware(hws []*api.Hardware) {
+	hwQu.mu.Lock()
+	hwQu.acquireHw(hws)
+	hwQu.mu.Unlock()
+}
+
+func (hwQu *hardwareCatalogue) acquireHw(hws []*api.Hardware) {
```

Comment:
I think acquire and release are the wrong way round. Acquire should return hardware to you as a consumer and release should put it back.

Reply:
good catch!